### ISSUE TITLE
MongoDB Atlas Integration

### ISSUE THREAD

**Issue by alexiantonino** [apache/superset#22787]
Hi There - I am the product manager for the  MongoDB Atlas SQL Interface. I have a few customers looking to use Superset with their MongoDB Atlas data.  We have a new JDBC Driver that uses our homegrown SQL dialect called mongosql. It is SQL-92 compatible with a few mongo specific functions added that help transform nested document data (flatten and unwind). 

I would love to explore the possibility of having a MongoDB Atlas integration for Superset. Here is a link to our JDBC Driver: https://search.maven.org/artifact/org.mongodb/mongodb-jdbc

And here is a link to our documentation for Atlas SQL: https://www.mongodb.com/docs/atlas/data-federation/query/query-with-sql/

And finally I also attached our mongosql reference guide, which will be published in our docs space within a few months.

[MongoSQL Reference Guide-4.pdf](https://github.com/apache/superset/files/10459709/MongoSQL.Reference.Guide-4.pdf)

Please let me know if there is anything else you need from me, or if we should have a meeting to discuss the possibility of an integration for MongoDB Atlas.

Thanks in advance,
Alexi Antonino
alexi.antonino@mongodb.com

**Comment by rusackas** [comment]
Hi! I believe the JDBC driver is not enough, and that to integrate with Superset, there will need to be a SQLAlchemy dialect. There's more information on this available [here](https://superset.apache.org/docs/databases/installing-database-drivers/). 

FWIW, we should probably move this from an Issue to a Github Q&A Discussion, but I'll leave it open here for a moment in order to assign the issue to a few folks who know more about this than I do, and can "check my math" and/or provide further assistance.

**Comment by alexiantonino** [comment]
Thanks so much. I will research the SQLAlchemy dialect info you've attached.

*{* name     : "Alexi Antonino",

  title    : "Product Manager",
  phone    : "248-835-6973",
  location : "Detroit, MI",
  linkedin : "alexiantonino <https://www.linkedin.com/in/alexiantonino/>"
*}*

**Comment by villebro** [comment]
@alexiantonino if MongoSQL is indeed SQL-92 compatible, creating a SQLAlchemy dialect for it should be very straight forward. Once that's done adding the necessary connector to Superset should be very straight forward, and we'd be happy to help get it integrated.

**Comment by srinify** [comment]
To add to this thread, I would read these 2 posts on what it takes to build a good database integration in Superset:

- https://preset.io/blog/building-database-connector/
- https://preset.io/blog/improving-apache-superset-integration-database-sqlalchemy/

**Comment by alexiantonino** [comment]
Thanks so much. I will look through the SQLAlchemy info and share it all
with my engineering team.  I appreciate the reply.

*{* name     : "Alexi Antonino",

  title    : "Product Manager",
  phone    : "248-835-6973",
  location : "Detroit, MI",
  linkedin : "alexiantonino <https://www.linkedin.com/in/alexiantonino/>"
*}*

**Comment by alexiantonino** [comment]
Thanks so much Srini! I will share this with my engineering team. I really
appreciate the response and info shared.

*{* name     : "Alexi Antonino",

  title    : "Product Manager",
  phone    : "248-835-6973",
  location : "Detroit, MI",
  linkedin : "alexiantonino <https://www.linkedin.com/in/alexiantonino/>"
*}*

**Comment by rusackas** [comment]
@alexiantonino fantastic! Holler if we can help. In the meantime, I'm wondering if we should close this as an issue (usually reserved for bugs) and either just await a PR when your eng team is ready, or convert this to an "Ideas" thread on the repo's Discussions, and start tracking it there. Happy to keep this open a bit while things are in the works, of course :D

**Comment by alexiantonino** [comment]
Yes please feel free to close this as an issue or convert it to an idea.
Thanks a bunch.

*{* name     : "Alexi Antonino",

  title    : "Product Manager",
  phone    : "248-835-6973",
  location : "Detroit, MI",
  linkedin : "alexiantonino <https://www.linkedin.com/in/alexiantonino/>"
*}*

**Comment by rusackas** [comment]
OK, closing for now. If you need any assistance along the way, we can still chat here or on a fresh Discussion. Looking forward to Atlas support in Superset, and we're more than happy to evangelize it when it lands!

**Comment by dwevedivaibhav** [comment]
Hi @alexiantonino or @rusackas 

I am currently exploring the integration of Superset with MongoDB Atlas SQL and would appreciate some guidance on the process.

Could you kindly provide instructions or examples on how to establish the connection? Specifically, I'm interested in knowing which connector is recommended for use in Superset.

Your assistance in this matter would be greatly appreciated.

Thank you in advance.

Thanks

**Comment by rusackas** [comment]
@dwevedivaibhav maybe this will help (even if it's a bit dated)? https://preset.io/blog/building-database-connector/
I'm not sure which driver would be used here... I'm not sure there's a SQLA dialect (yet) to lean on. @betodealmeida might have more advice here.

**Comment by rusackas** [comment]
Just to provide a small update, we had a call with some folks from Mongo, and are trying to find the right people that might be able to help steer toward a dbapi and complete this integration. Anyone finding this thread that wants to make Atlas work is still welcome to contribute to this effort.

**Comment by rusackas** [comment]
Just noting here that I found [this](https://github.com/apache/superset/issues/24352) thread, which led me to learning of [sqlalchemy-mongobi](https://pypi.org/project/sqlalchemy-mongobi/). I'm not sure how relevant or well-tested that is, but if there's a way to verify that it works, we should add that to our discussion/docs/etc as well!

**Comment by alexiantonino** [comment]
Thanks Evan -This is old, but it might help us to establish what to do/not do with the Atlas SQL driver. I'll read it and see if anything jumps out.

**Comment by alexiantonino** [comment]
@rusackas - I found this, the [SQLAlchemy JDBC api](https://pypi.org/project/sqlalchemy-jdbcapi/). This one seems like it is the most promising to potentially use with the [MongoDB Atlas SQL JDBC Driver](https://www.mongodb.com/try/download/jdbc-driver).  
Our SQL dialect is homegrown, called mongosql, but it is SQL-92 compatible. I mention this because our BI Connector runs with a mySQL dialect, and that is definitely supported in SQLAlchemy, but I didn't see anything more generic saying SQL-92, but the JDBC API seems to lean into the generic concept: ":In theory every database with a suitable JDBC driver should work."
I will research some more and keep you updated.

**Comment by kevin9foong** [comment]
hey @alexiantonino, just checking if there are any updates on connecting mongodb atlas SQL to superset? i'm hoping to set this up!

**Comment by villebro** [comment]
FYI, I think Trino has a [MongoDB connector](https://trino.io/docs/current/connector/mongodb.html), so using that would likely save a lot of trouble.
